lib/src/component/with_transition.dart

```diff
@@ -0,0 +1,410 @@
+        v2.TransitionPhase.HIDDEN: {},
+      }
+      ..addProps(v2.TransitionPropsMixin.defaultProps);
+    }, const []);
+    final props = defaultProps..addProps(_props);
+
+    assert(_hasSingleValidChild(props));
+
+    // ----- State / Hooks ----- //
+    final transitionNodeRef = useRef<Element>();
+    final transitionPhaseDerivedFromProps = props.isShown ? TransitionPhase.SHOWN : TransitionPhase.HIDDEN;
+    final transitionPhase = useState(transitionPhaseDerivedFromProps);
+
+    /// Begin showing the node unless the [currentPhase] is already shown or is in the process of showing.
+    void _handleShow(WithTransitionProps tProps, TransitionPhase currentPhase) {
+      if (_isOrWillBeShown(currentPhase)) return;
+
+      tProps.onWillShow?.call();
+      transitionPhase.set(_hasTransitionIn(tProps) ? TransitionPhase.PRE_SHOWING : TransitionPhase.SHOWN);
+    }
+
+    /// Begin hiding node unless the [currentPhase] is already hidden or is in the process of hiding.
+    void _handleHide(WithTransitionProps tProps, TransitionPhase currentPhase) {
+      if (_isOrWillBeHidden(currentPhase)) return;
+
+      tProps.onWillHide?.call();
+      transitionPhase.set(
+          _hasTransitionOut(tProps) ? TransitionPhase.HIDING : TransitionPhase.HIDDEN);
+    }
+
+    // getDerivedStateFromProps
+    if (props.isShown && _isOrWillBeHidden(transitionPhase.value)) {
+      _handleShow(props, transitionPhase.value);
+    } else if (!props.isShown && _isOrWillBeShown(transitionPhase.value)) {
+      _handleHide(props, transitionPhase.value);
+    }
+
+    // ----- transitionend event handling ----- //
+
+    /// Whether the overlay is not guaranteed to transition in response to the current
+    /// state change.
+    ///
+    /// _Stored as variable as workaround for not adding breaking change to [handleHiding] API._
+    ///
+    /// A transition may not always occur when the state moves from SHOWING to HIDING
+    /// if the PRE_SHOWING-->SHOWING-->HIDING transition happens back-to-back.
+    ///
+    /// Better to not always transition when the user is ninja-toggling a transitionable
+    /// component than to break state changes waiting for a transition that will never happen.
+    final _transitionNotGuaranteed = useRef(false);
+
+    /// Timer used to determine if a transition timeout has occurred.
+    final _transitionEndTimer = useRef<Timer>();
+
+    /// Stream for listening to `transitionend` events on the node.
+    final _endTransitionSubscription = useRef<StreamSubscription>();
+
+    void _cancelTransitionEventListener() {
+      _endTransitionSubscription.current?.cancel();
+      _endTransitionSubscription.current = null;
+    }
+
+    void _cancelTransitionEndTimer() {
+      _transitionEndTimer.current?.cancel();
+      _transitionEndTimer.current = null;
+    }
+
+    /// Listens for the next `transitionend` event and invokes a callback after
+    /// the event is dispatched.
+    void onNextTransitionEnd(Function() complete) {
```

Comment:
This references a number of props inside it (`props.transitionTimeout`, `_transitionInCount(props)`) that aren't declared in the calling `useLayoutEffect`

Reply:
Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8